src/advanced/JupyterNotebook/index.tsx

```diff
@@ -0,0 +1,327 @@
+        />
+    );
+};
+
+// -- renders a cell of type 'error'
+const JupyterNotebookOutputErrorCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    execution_count,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`In [${execution_count}]`}
+            bodyColumnContent={
+                <ReactMarkdown
+                    children={
+                        '**[' +
+                        output.evalue +
+                        ']**\n```\n ' +
+                        sourceLines(output.traceback).replace(
+                            /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g,
+                            ''
+                        ) +
+                        '\n```'
+                    }
+                />
+            }
+            bodyColumnStyle={{
+                padding: '0px 8px 0px 8px',
+                margin: '0px 0px',
+                border: '1px solid #AAA',
+                backgroundColor: '#FDD',
+            }}
+        />
+    );
+};
+
+// -- renders various data cell types (html, png, text)
+const JupyterNotebookOutputDataCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    execution_count,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`Out [${execution_count}]`}
+            bodyColumnContent={
+                <Container
+                    style={{
+                        marginBottom: '0px',
+                        boxShadow: 'none',
+                    }}
+                >
+                    {output.data['text/html'] ? (
+                        parse(sourceLines(output.data['text/html']), { trim: true })
+                    ) : output.data['image/png'] ? (
+                        <img
+                            src={'data:image/png;base64,' + output.data['image/png'].replace(/\n/g, '')}
+                            alt={'notebook diagram'}
+                        />
+                    ) : output.data['text/plain'] ? (
+                        <ReactMarkdown children={'```\n' + sourceLines(output.data['text/plain']) + '\n```'} />
+                    ) : (
+                        <b>unknown</b>
+                    )}
+                </Container>
+            }
+            bodyColumnStyle={{ padding: '0px 0px 0px 0px', margin: '0px 0px', border: '1px solid #AAA' }}
+        />
+    );
+};
+
+// -- helper class used in syntax highlighting the code
+const CodeBlock: CodeComponent = ({ inline = false, className, children }) => {
+    const match = /language-(\w+)/.exec(className || '');
+    let codeLanguage = 'javascript';
+    let langs = ['python', 'java', 'javascript', 'c++', 'typescript', 'objective-c', 'json'];
+
+    for (const lang of langs) {
+        if (match && match[1] && match[1].startsWith(lang)) {
+            codeLanguage = lang;
+        }
+    }
```

Comment:
Perhaps it's worth putting this in a `useMemo` so it's only recomputed when the `className` changes? :)

Can also simplify a little with `langs.find((lang) => match && match[1] && ...) || 'javascript'`

Reply:
useMemo definitely, not sure the find approach is clearer though